ci: Add feature to use different meson versions

nbyavuz opened the pull request:
We want to have a branch that uses a specific branch of meson for all tasks. Then, we can create a Cirrus cron "job" that tests Postgres againts different meson branches.

Set a different meson branch by using `MESON_REPO` and `MESON_BRANCH` environment variables. Use a bash script to install this specific meson branch and rebase current Postgres branch onto Postgres HEAD.

Fixes #81.

anarazel commented on .cirrus.yml:
```diff
@@ -95,6 +99,9 @@ task:
     chmod g+rwx /
 
+  install_meson_and_rebase_script: |
+    bash src/tools/ci/install_meson_and_rebase.sh ${MESON_REPO} ${MESON_BRANCH} linux
```
Could we use ${CIRRUS_OS} instead of needing to write the os ourselves?

anarazel replied:
Also, I think the script already would get $MESON_BRANCH etc as an environment variable, so we don't need to explicitly pass it. Normally explicit is better, but that way we don't need to deal with differences between shells etc?

anarazel commented on .cirrus.yml:
```diff
@@ -360,6 +370,9 @@ task:
 
+      install_meson_and_rebase_script: |
+        bash src/tools/ci/install_meson_and_rebase.sh ${MESON_REPO} ${MESON_BRANCH} linux
+
```
I think it's probably worth just doing this for all tasks, rather than just this one in the matrix.

I'd really like if we could put the invocation into a single place instead of needing to repeat it in many places. But it's not exactly obvious how. Looks like I should have made all tasks in postgres' .cirrus inline the same yaml block, to make it easier to do stuff like this.

I was wondering if we could use https://cirrus-ci.org/guide/programming-tasks/ to make this easier, but I couldn't find a way.

nbyavuz replied:
Yes, copying same thing on multiple places doesn't seem good but I couldn't find a way too.

anarazel replied:
I wonder, not as part of this PR, whether we should something like
```yaml
task_prep: &task_prep
  # add scripts to be executed at the start of all tasks here

task_finish: &task_finish
  # add scripts to be executed at the end of all tasks here
```

I would like to filter the set of artifacts uploaded in case of failures, which would be a lot easier if we had something like this...

nbyavuz replied:
Yes, I was thinking if we could ask for a feature request for something like this to cirrus?

anarazel commented on src/tools/ci/install_meson_and_rebase.sh:
```diff
@@ -0,0 +1,86 @@
+fi
+
+case $CURRENT_OS in
+    freebsd|linux|macos|windows|mingw)
```
It'd be good to handle net/openbsd here as well.

nbyavuz replied:
Should I target meson branch instead of meson-upstream-master? There are no net/openbsd tasks on meson-upstream-master.

nbyavuz replied:
I am planning to work on other comments after your answer.

anarazel replied:
> Should I target meson branch instead of meson-upstream-master? There are no net/openbsd tasks on meson-upstream-master.

Hm. I think we should start to upstream the net/openbsd tasks separately. 

It's probably not worth using my meson branch as base, it ends up having conflicts way more often...

anarazel commented on src/tools/ci/install_meson_and_rebase.sh:
```diff
@@ -0,0 +1,86 @@
+    chown -R postgres:postgres .[^.]*
+fi
+
+# pip is not installed as a default on some OSes. First install
```
How about we instead change the image definitions so that pip is included? That's also useful for patches to postgres that need other dependencies / utilities.

anarazel commented on src/tools/ci/install_meson_and_rebase.sh:
```diff
@@ -0,0 +1,86 @@
+            git config user.name 'Postgres CI'
+            # There are file permission issues on Windows. Set filemode to
+            # false for windows, then reset
+            git config core.filemode false
```
What are these issues?

nbyavuz replied:
Chmod issues, for example:
```shell
diff --git a/config/install-sh b/config/install-sh
old mode 100755
new mode 100644
```

anarazel replied:
Maybe change the comment to "windows looses the executable bit, causing an unnecessary diff".

And as I mentioned, I think this would look a lot better if just that  part were OS dependent.

anarazel commented on src/tools/ci/install_meson_and_rebase.sh:
```diff
@@ -0,0 +1,86 @@
+            pip3 install git+${MESON_REPO}@${MESON_BRANCH}
+        ;;
+        windows)
+            pip install git+${MESON_REPO}@${MESON_BRANCH}
```
How about the OS dependent bit just defines $PIP and then we have one pip line to install below? Seems not implausible we'd want another pip invocation somewhere.  Alternatively we could make sure that the OS images define $PIP?

nbyavuz replied:
> Alternatively we could make sure that the OS images define $PIP?

I tried this but I couldn't make it work. Also, we can't define variables on MacOS so I didn't spend much time on it.

anarazel commented on src/tools/ci/install_meson_and_rebase.sh:
```diff
@@ -0,0 +1,86 @@
+            # false for windows, then reset
+            git config core.filemode false
+            git reset --hard
+            git remote add default-postgres https://github.com/postgres/postgres.git
```
The majority of the case "body" is duplicated - why don't we just do the core.filemode bit in case?

nbyavuz commented:
I pushed this with a commit to run on 'pg-ci-images-dev' project since https://github.com/anarazel/pg-vm-images/pull/70 didn't run on 'pg-ci-images' yet. After https://github.com/anarazel/pg-vm-images/pull/70 is merged, I will push this without  "run on 'pg-ci-images-dev'" commit.

anarazel commented:
I just tested it for the new meson version about to be released - very useful!

I got one failure that i'm somewhat confused by: https://cirrus-ci.com/task/6190199437262848?logs=configure#L8
that doesn't seem related to the new meson version, but doesn't happen otherwise?

An improvement for this would be to list which commit we rebased onto:
https://cirrus-ci.com/task/6190199437262848?logs=install_meson_and_rebase#L46
doesn't allow to figure that out:
```
[21:31:59.899] + git remote add default-postgres https://github.com/postgres/postgres.git
[21:31:59.930] + git fetch default-postgres master
[21:32:04.399] From https://github.com/postgres/postgres
[21:32:04.399]  * branch              master     -> FETCH_HEAD
[21:32:04.399]  * [new branch]        master     -> default-postgres/master
[21:32:04.415] + git rebase --no-verify default-postgres/master
[21:32:05.879] Rebasing (1/1)
Successfully rebased and updated detached HEAD.
```
Something like
```
echo Rebasing onto: $(git show --no-patch --abbrev-commit  --pretty=oneline default-postgres/master)
```
maybe?

nbyavuz commented:
> I got one failure that i'm somewhat confused by: https://cirrus-ci.com/task/6190199437262848?logs=configure#L8 that doesn't seem related to the new meson version, but doesn't happen otherwise?

I found the cause and fixed it. It seems rebase part changing file endings to CLRF on Windows, I added `git config core.autocrlf false`. 

Also, I realized that CI only collects `*.diffs` files but sometimes diff logs are written in `*.diff` files (for example pg_bsd_indent). It can be fixed by editing on_failure parts but I couldn't decide between editing on_failure parts and changing *.diff to *.diffs.

> Something like
> 
> ```
> echo Rebasing onto: $(git show --no-patch --abbrev-commit  --pretty=oneline default-postgres/master)
> ```

Added.

tristan957 commented on src/tools/ci/install_meson_and_rebase.sh:
```diff
@@ -0,0 +1,66 @@
+run_rebase_commands() {
+    git config user.email 'postgres-ci@example.com'
+    git config user.name 'Postgres CI'
+    # windows looses the executable bit, causing an unnecessary diff
```
s/looses/loses

anarazel commented:
> > I got one failure that i'm somewhat confused by: https://cirrus-ci.com/task/6190199437262848?logs=configure#L8 that doesn't seem related to the new meson version, but doesn't happen otherwise?
> 
> I found the cause and fixed it. It seems rebase part changing file endings to CLRF on Windows, I added `git config core.autocrlf false`.

Great.

> Also, I realized that CI only collects `*.diffs` files but sometimes diff logs are written in `*.diff` files (for example pg_bsd_indent). It can be fixed by editing on_failure parts but I couldn't decide between editing on_failure parts and changing *.diff to *.diffs.

Heh, yea, I noticed that too and even wrote an email about it:
https://www.postgresql.org/message-id/20230711233307.hu4wetabjm5f7ver%40awork3.anarazel.de
and am planning to rename the file.

> > Something like
> > ```
> > echo Rebasing onto: $(git show --no-patch --abbrev-commit  --pretty=oneline default-postgres/master)
> > ```
> 
> Added.

Thanks!

anarazel commented on src/tools/ci/install_meson_and_rebase.sh:
```diff
@@ -0,0 +1,66 @@
+# Rebase current branch onto Postgres HEAD
+rebase_onto_postgres () {
+    echo "Rebasing onto: $(git show --no-patch --abbrev-commit  --pretty=oneline default-postgres/master)"
+    # freebsd and linux run commands as postgres user
```
Is that actually required? It's true that they run configure / build as postgres, but that shouldn't be affected by source code permissions?

anarazel commented on src/tools/ci/install_meson_and_rebase.sh:
```diff
@@ -0,0 +1,66 @@
+    git config core.filemode false
+    # windows changes file endings to crlf, causing test failures
+    git config core.autocrlf false
+    git reset --hard
```
Why is the reset --hard needed?

anarazel commented on src/tools/ci/install_meson_and_rebase.sh:
```diff
@@ -0,0 +1,66 @@
+esac
+
+# After repartition, hidden files are not copied. Copy them to
+# working dir
```
I'd rather fix that in the repartition step?

anarazel commented on src/tools/ci/install_meson_and_rebase.sh:
```diff
@@ -0,0 +1,66 @@
+            PYTHON=python
+        ;;
+        *)
+            PYTHON=python3
```
I assume there's no 'python3' on windows? If not, perhaps we should just add an alias to the image? That's annoying in other places too...
